crates/multi-transaction-payment/src/tests.rs

```diff
@@ -0,0 +1,417 @@
+            let foreign_currency = GetForeignCurrencyId::get();
+
+            setup_dex(native_currency, foreign_currency);
+            let call = RuntimeCall::MultiTransactionPayment(crate::Call::with_fee_swap_path {
+                amount_in_max: u128::MAX,
+                path: vec![foreign_currency, native_currency],
+                call: Box::new(RuntimeCall::Testing(testing_helpers::Call::weighted {
+                    expected: Weight::from_ref_time(expected_weight),
+                    actual_weight: actual_weight.map(|x| Weight::from_ref_time(x)),
+                    err,
+                    pays_fee,
+                })),
+            });
+            dry_run_given_call(who, native_currency, foreign_currency, pays_fee, call, tip)
+        })
+    }
+
+    pub(super) fn validate_swapping_payment(
+        expected_weight: u64,
+        amount_in_max: u128,
+        swap_path: (CurrencyId, CurrencyId),
+        initial_fee_balance: u128,
+    ) -> TransactionValidity {
+        dry_run(|| {
+            let who = 1;
+            let native_currency = GetNativeCurrencyId::get();
+            let foreign_currency = GetForeignCurrencyId::get();
+
+            setup_dex(native_currency, foreign_currency);
+            let call = RuntimeCall::MultiTransactionPayment(crate::Call::with_fee_swap_path {
+                amount_in_max: amount_in_max,
+                path: vec![foreign_currency, native_currency],
+                call: Box::new(RuntimeCall::Testing(testing_helpers::Call::weighted {
+                    expected: Weight::from_ref_time(expected_weight),
+                    actual_weight: None,
+                    err: false,
+                    pays_fee: Pays::Yes,
+                })),
+            });
+            assert_ok!(Tokens::set_balance(
+                RuntimeOrigin::root(),
+                who,
+                swap_path.0,
+                initial_fee_balance,
+                0
+            ));
+
+            let payment_extension = pallet_transaction_payment::ChargeTransactionPayment::<Test>::from(0);
+            let info = call.get_dispatch_info();
+            let len = 1231;
+
+            dry_run(|| payment_extension.validate(&who, &call, &info, len))
+        })
+    }
+
+    pub(super) fn dry_run_swapped(
+        expected_weight: u64,
+        actual_weight: Option<u64>,
+        err: bool,
+        pays_fee: Pays,
+    ) -> BalanceChange {
+        dry_run_swapped_with_tip(expected_weight, actual_weight, err, pays_fee, 0)
+    }
+}
+
+#[test]
+/// basic test that payment in native fee continues to work as normal
+fn pay_in_native_fee_succeeds() {
+    new_test_ext().execute_with(|| {
+        // <Payment as SignedExtension>::pre_dispatch();
+        let payment_extension = pallet_transaction_payment::ChargeTransactionPayment::<Test>::from(1);
+        let who = 1;
+        let call = RuntimeCall::System(frame_system::Call::remark { remark: vec![] });
```

Comment:
Similar to this, is it worth testing a call wrapped in `with_fee_swap_path` where the path has one or two (duplicate) entries of the native currency?

Reply:
I don't think so - all this pallet cares about is that the path ends in the native token. If the dex can't handle paths with duplicate entries, it should have unit tests for that inside the dex pallet itself